price-estimator/src/filter.rs

```diff
@@ -0,0 +1,218 @@
+        let mut split = s.split('-');
+        let mut next_token_id = || -> Result<u16, ParseTokenPairError> {
+            let token_string = split
+                .next()
+                .ok_or(ParseTokenPairError::WrongNumberOfTokens)?;
+            token_string.parse().map_err(From::from)
+        };
+        let buy_token_id = next_token_id()?;
+        let sell_token_id = next_token_id()?;
+        if split.next().is_some() {
+            return Err(ParseTokenPairError::WrongNumberOfTokens);
+        }
+        Ok(Self {
+            buy_token_id,
+            sell_token_id,
+        })
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct JsonResult {
+    #[serde(with = "display_fromstr")]
+    base_token_id: u16,
+    #[serde(with = "display_fromstr")]
+    quote_token_id: u16,
+    #[serde(with = "display_fromstr")]
+    buy_amount_in_base: u128,
+    #[serde(with = "display_fromstr")]
+    sell_amount_in_quote: u128,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::future::FutureExt as _;
+    use serde_json::Value;
+
+    #[test]
+    fn estimated_buy_amount_ok() {
+        let (token_pair, volume) = warp::test::request()
+            .path("/markets/0-65535/estimated-buy-amount/1")
+            .filter(&estimated_buy_amount_filter())
+            .now_or_never()
+            .unwrap()
+            .unwrap();
+        assert_eq!(token_pair.buy_token_id, 0);
+        assert_eq!(token_pair.sell_token_id, 65535);
+        assert_eq!(volume, 1);
+    }
+
+    #[test]
+    fn estimated_buy_amount_too_few_tokens() {
+        for path in &[
+            "/markets//estimated-buy-amount/1",
+            "/markets/0/estimated-buy-amount/1",
+        ] {
+            assert!(warp::test::request()
+                .path(path)
+                .filter(&estimated_buy_amount_filter())
+                .now_or_never()
+                .unwrap()
+                .is_err());
+        }
+    }
+
+    #[test]
+    fn estimated_buy_amount_too_many_tokens() {
+        for path in &[
+            "/markets/0-1-2/estimated-buy-amount/1",
+            "/markets/0-1-asdf/estimated-buy-amount/1",
+            "/markets/0-1-2-3/estimated-buy-amount/1",
+            "/markets/0-1-/estimated-buy-amount/1",
+        ] {
+            assert!(warp::test::request()
+                .path(path)
+                .filter(&estimated_buy_amount_filter())
+                .now_or_never()
+                .unwrap()
+                .is_err());
+        }
+    }
+
+    #[test]
+    fn estimated_buy_amount_no_volume() {
+        for path in &[
+            "/markets/0-1/estimated-buy-amount/",
+            "/markets/0-1/estimated-buy-amount/asdf",
+        ] {
+            assert!(warp::test::request()
+                .path(path)
+                .filter(&estimated_buy_amount_filter())
+                .now_or_never()
+                .unwrap()
+                .is_err());
+        }
```

Comment:
This looks more like an error test rather than "no volume"

Reply:
What it's trying to test is to see what happens when the sell_amount_in_quote is missing. I called it "volume" here instead but I'll rename it.